src/solvers/equation_solver.cpp

```diff
@@ -13,39 +13,113 @@ namespace serac {
+  } else {
+    lin_solver_ = buildIterativeLinearSolver(comm, lin_params);
+  }
+
+  if (nonlin_params) {
+    nonlin_solver_ = buildNewtonSolver(comm, *nonlin_params, linearSolver());
+  }
+}
+
+std::unique_ptr<mfem::IterativeSolver> EquationSolver::buildIterativeLinearSolver(
+    MPI_Comm comm, const LinearSolverParameters& lin_params)
+{
+  std::unique_ptr<mfem::IterativeSolver> iter_lin_solver;
+
   switch (lin_params.lin_solver) {
     case LinearSolver::CG:
-      iter_lin_solver_ = std::make_unique<mfem::CGSolver>(comm);
+      iter_lin_solver = std::make_unique<mfem::CGSolver>(comm);
       break;
     case LinearSolver::GMRES:
-      iter_lin_solver_ = std::make_unique<mfem::GMRESSolver>(comm);
+      iter_lin_solver = std::make_unique<mfem::GMRESSolver>(comm);
       break;
     case LinearSolver::MINRES:
-      iter_lin_solver_ = std::make_unique<mfem::MINRESSolver>(comm);
+      iter_lin_solver = std::make_unique<mfem::MINRESSolver>(comm);
       break;
     default:
       SLIC_ERROR("Linear solver type not recognized.");
       exitGracefully(true);
   }
-  iter_lin_solver_->SetRelTol(lin_params.rel_tol);
-  iter_lin_solver_->SetAbsTol(lin_params.abs_tol);
-  iter_lin_solver_->SetMaxIter(lin_params.max_iter);
-  iter_lin_solver_->SetPrintLevel(lin_params.print_level);
 
-  if (nonlin_params) {
-    auto newton_solver = std::make_unique<mfem::NewtonSolver>(comm);
-    newton_solver->SetSolver(*iter_lin_solver_);
-    newton_solver->SetRelTol(nonlin_params->rel_tol);
-    newton_solver->SetAbsTol(nonlin_params->abs_tol);
-    newton_solver->SetMaxIter(nonlin_params->max_iter);
-    newton_solver->SetPrintLevel(nonlin_params->print_level);
-    nonlin_solver_ = std::move(newton_solver);
+  iter_lin_solver->SetRelTol(lin_params.rel_tol);
+  iter_lin_solver->SetAbsTol(lin_params.abs_tol);
+  iter_lin_solver->SetMaxIter(lin_params.max_iter);
+  iter_lin_solver->SetPrintLevel(lin_params.print_level);
+
+  return iter_lin_solver;
+}
+
+std::unique_ptr<mfem::NewtonSolver> EquationSolver::buildNewtonSolver(MPI_Comm                         comm,
+                                                                      const NonlinearSolverParameters& nonlin_params,
+                                                                      mfem::Solver&                    lin_solver)
+{
+  auto newton_solver = std::make_unique<mfem::NewtonSolver>(comm);
+  newton_solver->SetSolver(lin_solver);
+  newton_solver->SetRelTol(nonlin_params.rel_tol);
+  newton_solver->SetAbsTol(nonlin_params.abs_tol);
+  newton_solver->SetMaxIter(nonlin_params.max_iter);
+  newton_solver->SetPrintLevel(nonlin_params.print_level);
+  return newton_solver;
+}
+
+void EquationSolver::SetOperator(const mfem::Operator& op)
+{
+  if (nonlin_solver_) {
+    if (std::holds_alternative<std::unique_ptr<mfem::SuperLUSolver>>(lin_solver_)) {
+      superlu_wrapper_ = std::make_unique<SuperLUNonlinearOperatorWrapper>(op);
```

Comment:
This is called at every timestep. Will calling `make_unique` instead of `reset` leak memory?

Reply:
My googling shows it doesn't. TIL! There's probably a bunch of `reset`s that we don't need hanging around in the code due to my misunderstanding....